File: competitive-programming/src/edu/algos/li14_linkedlist/_L203_RemoveLinkedListElements.java
package edu.algos.li14_linkedlist;

import edu.algos.li00_model.ListNode;

public class _L203_RemoveLinkedListElements {

	public static void main(String[] args) {
		_L203_RemoveLinkedListElements obj = new _L203_RemoveLinkedListElements();
		
		_G01_SingleLinkedList list = new _G01_SingleLinkedList();
		list.insertAtEnd(2);
		list.insertAtEnd(4);
		list.insertAtEnd(3);
		list.insertAtEnd(2);
		
		list.head = obj.removeElements(list.head, 2);
		list.print();		
	}

	/*
	 * Problem: https://leetcode.com/problems/remove-linked-list-elements/
	 * Idea: We might have to remove head also. To simplify that add dummy node.
	 * If curr.next is val, then adjust curr.next = curr.next.next.
	 * 
	 * */
	public ListNode removeElements(ListNode head, int val) {
        ListNode start = new ListNode();
        start.next = head ; 

        ListNode curr = start;
        while(curr != null && curr.next != null){
            if(curr.next.val == val)
                curr.next = curr.next.next;
            else curr = curr.next;
        }
        
        return start.next;
    }
}

File: dsa/patterns/09_linkedlist/python/l203_remove_linked_list_elements.py
"""
Problem: https://leetcode.com/problems/remove-linked-list-elements/

Idea:
We might have to remove head also.
To simplify that add start node.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def remove_elements(head: ListNode, val: int) -> ListNode:
    start = ListNode(0)
    start.next = head

    curr = start
    while curr and curr.next:
        if curr.next.val == val:
            curr.next = curr.next.next
        else:
            curr = curr.next

    return start.next


# Example usage
if __name__ == "__main__":
    # 2 -> 4 -> 3 -> 2, remove 2
    n1 = ListNode(2)
    n2 = ListNode(4)
    n3 = ListNode(3)
    n4 = ListNode(2)

    n1.next = n2
    n2.next = n3
    n3.next = n4

    head = remove_elements(n1, 2)

    curr = head
    while curr:
        print(curr.val, end=" ")
        curr = curr.next
    # Output: 4 3
